econml/dml.py

```diff
@@ -50,12 +51,237 @@
                              TreatmentExpansionMixin, StatsModelsCateEstimatorMixin,
                              DebiasedLassoCateEstimatorMixin)
-from .inference import StatsModelsInference
+from .inference import StatsModelsInference, GenericSingleTreatmentModelFinalInference
 from ._rlearner import _RLearner
 from .sklearn_extensions.model_selection import WeightedStratifiedKFold
 
 
-class DMLCateEstimator(_RLearner):
+class _FirstStageWrapper:
+    def __init__(self, model, is_Y, featurizer, linear_first_stages, discrete_treatment):
+        self._model = clone(model, safe=False)
+        self._featurizer = clone(featurizer, safe=False)
+        self._is_Y = is_Y
+        self._linear_first_stages = linear_first_stages
+        self._discrete_treatment = discrete_treatment
+
+    def _combine(self, X, W, n_samples, fitting=True):
+        if X is None:
+            # if both X and W are None, just return a column of ones
+            return (W if W is not None else np.ones((n_samples, 1)))
+        XW = hstack([X, W]) if W is not None else X
+        if self._is_Y and self._linear_first_stages:
+            if self._featurizer is None:
+                F = X
+            else:
+                F = self._featurizer.fit_transform(X) if fitting else self._featurizer.transform(X)
+            return cross_product(XW, hstack([np.ones((shape(XW)[0], 1)), F]))
+        else:
+            return XW
+
+    def fit(self, X, W, Target, sample_weight=None):
+        if (not self._is_Y) and self._discrete_treatment:
+            # In this case, the Target is the one-hot-encoding of the treatment variable
+            # We need to go back to the label representation of the one-hot so as to call
+            # the classifier.
+            if np.any(np.all(Target == 0, axis=0)) or (not np.any(np.all(Target == 0, axis=1))):
+                raise AttributeError("Provided crossfit folds contain training splits that " +
+                                     "don't contain all treatments")
+            Target = inverse_onehot(Target)
+
+        if sample_weight is not None:
+            self._model.fit(self._combine(X, W, Target.shape[0]), Target, sample_weight=sample_weight)
+        else:
+            self._model.fit(self._combine(X, W, Target.shape[0]), Target)
+
+    def predict(self, X, W):
+        n_samples = X.shape[0] if X is not None else (W.shape[0] if W is not None else 1)
+        if (not self._is_Y) and self._discrete_treatment:
+            return self._model.predict_proba(self._combine(X, W, n_samples, fitting=False))[:, 1:]
+        else:
+            return self._model.predict(self._combine(X, W, n_samples, fitting=False))
+
+
+class _FinalWrapper:
+    def __init__(self, model_final, fit_cate_intercept, featurizer, use_weight_trick):
+        self._model = clone(model_final, safe=False)
+        self._use_weight_trick = use_weight_trick
+        self._original_featurizer = clone(featurizer, safe=False)
+        if self._use_weight_trick:
```

Comment:
The fit_cate_intercept is only a property that starts from the DMLCateEstimator parent node. It's not at the _RLearner level and the NonParamDMLCateEstimator inherits directyly from the _RLearner. Also here the cate intercept is the intercept of the final model (if it happens to be linear), but we don't expose any API to access these properties (which is the right thing). If you want a linear model, then use the linear DML. I want to let the featurizer in there, just so that someone doesn't need to write the pipeline every time. Also it's even worse: they will need to write a wrapper of a Pipeline, that accepts weights as wrapping an estimator that accepts weights with a pipeline doesn't immediately expose the weights. @moprescu  has written a `fit_weighted_pipeline` in some utilities, but I want to avoid this complication here.

Reply:
Okay, in that case I'd assert `not (use_weight_trick and fit_cate_intercept)` at the top because we don't ever expect this class to be used that way.

Your pipeline reasoning makes sense.

Reply:
Added an extra condition. Though the final wrapper at init time, when use_weight_trick is True it automatically sets fit_cate_intercept to False.